### PR TITLE
Avoid macros for CUDA lock arrays

### PR DESCRIPTION
**Pull request by masterleinad**
Trying the suggestion is #2235.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
We should of course critically examine if this works.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Submit PR on Desul side

**Comment by dalg24** [comment]
Ping

**Comment by masterleinad** [comment]
> Submit PR on Desul side

See https://github.com/desul/desul/pull/64.

**Comment by dalg24** [comment]
Please fix conflicts

**Comment by masterleinad** [comment]
I update with the changes in `desul`.

**Review by dalg24 — COMMENTED** [review]
Do you know why we have both the Kokkos and the Desul lock arrays?

**Comment by masterleinad** [comment]
> Do you know why we have both the Kokkos and the Desul lock arrays?

`g_device_cuda_lock_arrays` is also used in `Kokkos_Cuda_Parallel_Team.hpp`. 
I guess we could discuss changing that.

File: core/src/Cuda/Kokkos_Cuda_Locks.hpp [hunk 1]
```diff
@@ -151,45 +151,50 @@ namespace {
...
+#else
+static
+#endif
+    void
+    copy_cuda_lock_arrays_to_device() {
+  if (::Kokkos::Impl::lock_array_copied == 0) {
+    KOKKOS_IMPL_CUDA_SAFE_CALL(
+        cudaMemcpyToSymbol(Kokkos::Impl::g_device_cuda_lock_arrays,
+                           &Kokkos::Impl::g_host_cuda_lock_arrays,
+                           sizeof(Kokkos::Impl::CudaLockArrays)));
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Locks.hpp]: See [f7d1303](https://github.com/kokkos/kokkos/pull/4682/commits/f7d1303e2ff25db9b085b1466f5a3f9b3869a37a).

File: core/src/Cuda/Kokkos_Cuda_Locks.hpp [hunk 2]
```diff
@@ -151,45 +149,49 @@ namespace {
...
+                                                  sizeof(CudaLockArrays)));
   }
+  lock_array_copied = 1;
+}
 
 #ifndef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
 
 #ifdef KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE
-#define KOKKOS_ENSURE_CUDA_LOCK_ARRAYS_ON_DEVICE()
+void inline ensure_cuda_lock_arrays_on_device() {}
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Locks.hpp:171]:
```suggestion
inline void ensure_cuda_lock_arrays_on_device() {}
```

File: core/src/Cuda/Kokkos_Cuda_Locks.hpp [hunk 3]
```diff
@@ -151,45 +149,49 @@ namespace {
...
 #else
-#define KOKKOS_ENSURE_CUDA_LOCK_ARRAYS_ON_DEVICE() \
-  KOKKOS_COPY_CUDA_LOCK_ARRAYS_TO_DEVICE()
+void static inline ensure_cuda_lock_arrays_on_device() {
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Locks.hpp:173]:
```suggestion
inline static void ensure_cuda_lock_arrays_on_device() {
```

File: core/src/Cuda/Kokkos_Cuda_Locks.hpp [hunk 4]
```diff
@@ -151,45 +149,49 @@ namespace {
...
 
 #ifdef KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE
-#define KOKKOS_ENSURE_CUDA_LOCK_ARRAYS_ON_DEVICE()
+void inline ensure_cuda_lock_arrays_on_device() {}
 #else
 // Still Need COPY_CUDA_LOCK_ARRAYS for team scratch etc.
-#define KOKKOS_ENSURE_CUDA_LOCK_ARRAYS_ON_DEVICE() \
-  KOKKOS_COPY_CUDA_LOCK_ARRAYS_TO_DEVICE()         \
-  DESUL_ENSURE_CUDA_LOCK_ARRAYS_ON_DEVICE()
+void static inline ensure_cuda_lock_arrays_on_device() {
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Locks.hpp:184]:
```suggestion
inline static void ensure_cuda_lock_arrays_on_device() {
```

**Review by dalg24 — DISMISSED** [review]

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
Does not pass CI

File: core/src/Cuda/Kokkos_Cuda_Locks.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Locks.hpp]: 0a3e71b84c07d0677ad8f6d9d62e7f1017f757af

File: core/src/Cuda/Kokkos_Cuda_Locks.hpp (hunk 3 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Locks.hpp]: 0a3e71b84c07d0677ad8f6d9d62e7f1017f757af

**Review by dalg24 — APPROVED** [review]

**Pull request merged**